[SPARK-17471][ML] Add compressed method to ML matrices

sethah opened the pull request:
## What changes were proposed in this pull request?

This patch adds a `compressed` method to ML `Matrix` class, which returns the minimal storage representation of the matrix - either sparse or dense. Because the space occupied by a sparse matrix is dependent upon its layout (i.e. column major or row major), this method must consider both cases. It may also be useful to force the layout to be column or row major beforehand, so an overload is added which takes in a `columnMajor: Boolean` parameter.

The compressed implementation relies upon two new abstract methods `toDense(columnMajor: Boolean)` and `toSparse(columnMajor: Boolean)`, similar to the compressed method implemented in the `Vector` class. These methods also allow the layout of the resulting matrix to be specified via the `columnMajor` parameter. More detail on the new methods is given below.
## How was this patch tested?

Added many new unit tests
## New methods (summary, not exhaustive list)

**Matrix trait**
- `private[ml] def toDenseMatrix(columnMajor: Boolean): DenseMatrix` (abstract) - converts the matrix (either sparse or dense) to dense format
- `private[ml] def toSparseMatrix(columnMajor: Boolean): SparseMatrix` (abstract) -  converts the matrix (either sparse or dense) to sparse format
- `def toDense: DenseMatrix = toDense(true)`  - converts the matrix (either sparse or dense) to dense format in column major layout
- `def toSparse: SparseMatrix = toSparse(true)` -  converts the matrix (either sparse or dense) to sparse format in column major layout
- `def compressed: Matrix` - finds the minimum space representation of this matrix, considering both column and row major layouts, and converts it
- `def compressed(columnMajor: Boolean): Matrix` - finds the minimum space representation of this matrix considering only column OR row major, and converts it

**DenseMatrix class**
- `private[ml] def toDenseMatrix(columnMajor: Boolean): DenseMatrix` - converts the dense matrix to a dense matrix, optionally changing the layout (data is NOT duplicated if the layouts are the same)
- `private[ml] def toSparseMatrix(columnMajor: Boolean): SparseMatrix` - converts the dense matrix to sparse matrix, using the specified layout

**SparseMatrix class**
- `private[ml] def toDenseMatrix(columnMajor: Boolean): DenseMatrix` - converts the sparse matrix to a dense matrix, using the specified layout
- `private[ml] def toSparseMatrix(columnMajors: Boolean): SparseMatrix` - converts the sparse matrix to sparse matrix. If the sparse matrix contains any explicit zeros, they are removed. If the layout requested does not match the current layout, data is copied to a new representation. If the layouts match and no explicit zeros exist, the current matrix is returned.

sethah committed:
first commit

sethah committed:
start to add tests

sethah committed:
sparse to sparse stuff

sethah committed:
improve test cases and cleanup

sethah committed:
adding some helper methods and shoring up test cases

sethah committed:
cleanup

sethah committed:
minor cleanup

SparkQA commented:
**[Test build #67527 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67527/consoleFull)** for PR 15628 at commit [`a51e217`](https://github.com/apache/spark/commit/a51e2173089cf79781b0d9a37492a4c4b4080881).

sethah committed:
minor syntactical changes

SparkQA commented:
**[Test build #67528 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67528/consoleFull)** for PR 15628 at commit [`5600d14`](https://github.com/apache/spark/commit/5600d143ba3a06af9b2b1829a2b02f7e537b3814).

sethah commented:
cc @dbtsai @yanboliang 

**Note for reviewers**: The driving need behind this change is that we now have multiclass logistic regression which represents coefficients as a matrix. When we use L1 regularization it can be more efficient to store the coefficient matrix as sparse since there can be a lot of zero elements. We will incorporate this compression in a follow-up PR

SparkQA commented:
**[Test build #67527 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67527/consoleFull)** for PR 15628 at commit [`a51e217`](https://github.com/apache/spark/commit/a51e2173089cf79781b0d9a37492a4c4b4080881).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67527/
Test FAILed.

SparkQA commented:
**[Test build #67528 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67528/consoleFull)** for PR 15628 at commit [`5600d14`](https://github.com/apache/spark/commit/5600d143ba3a06af9b2b1829a2b02f7e537b3814).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67528/
Test FAILed.

sethah commented:
I'm not sure I understand why there are MiMa failures. It should not be a problem that the new methods only exist in the current version AFAIK.

sethah committed:
add overrides, rearrange code, minor cleanups

SparkQA commented:
**[Test build #67819 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67819/consoleFull)** for PR 15628 at commit [`b5277c9`](https://github.com/apache/spark/commit/b5277c9bffef72b207f3d79f11b7bb01661de9e1).

sethah committed:
move toDense and toSparse to trait

SparkQA commented:
**[Test build #67820 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67820/consoleFull)** for PR 15628 at commit [`a931fb9`](https://github.com/apache/spark/commit/a931fb9e0d13a25341865508c829c0506387ecec).

sethah commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -153,6 +153,86 @@ sealed trait Matrix extends Serializable {
+   * @param columnMajor Whether the values of the resulting dense matrix should be in column major
+   *                    or row major order. If `false`, resulting matrix will be row major.
+   */
+  private [ml] def toDenseMatrix(columnMajor: Boolean): DenseMatrix
```
minor: space

sethah commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -153,6 +153,86 @@ sealed trait Matrix extends Serializable {
+   * @param columnMajor Whether the values of the resulting sparse matrix should be in column major
+   *                    or row major order. If `false`, resulting matrix will be row major.
+   */
+  private[ml] def toSparseMatrix(columnMajor: Boolean): SparseMatrix
```
A bit of explanation: I made this a private method with a different name because `toDense: DenseMatrix` and `toSparse: SparseMatrix` should be implemented in the trait, not in the subclasses. But we can't just put them here and use overloading, because we will get ambiguous reference compile errors. So, we implement them here and make this private with a different name to avoid this. I appreciate feedback on this approach - it feels a bit awkward.

sethah replied:
Should we just leave `toDenseMatrix` private and have `toDense` always use colMajor = true? I think that's ok to do for now.

SparkQA commented:
**[Test build #67819 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67819/consoleFull)** for PR 15628 at commit [`b5277c9`](https://github.com/apache/spark/commit/b5277c9bffef72b207f3d79f11b7bb01661de9e1).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67819/
Test FAILed.

SparkQA commented:
**[Test build #67820 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/67820/consoleFull)** for PR 15628 at commit [`a931fb9`](https://github.com/apache/spark/commit/a931fb9e0d13a25341865508c829c0506387ecec).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/67820/
Test FAILed.

WeichenXu123 commented:
@sethah I think this is a good idea, I am going to add the same method to Vector, since high dimension data training on LIR/LOR with L1 reg, compressed coeffs storage will make the algo broadcast coeffs vector faster, how do you think about it ?

sethah commented:
What are you going to add? `compressed` already exists on vectors...

WeichenXu123 commented:
@sethah Oh..I miss it..

MLnick commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -1076,4 +1240,15 @@ object Matrices {
+
+  private[ml] def getSparseSize(numActives: Long, numPtrs: Long): Long = {
+    // 8 * values.length + 4 * rowIndices.length + 4 * colPtrs.length + 8 + 8 + 1
+    12L * numActives + 4L * numPtrs + 17L
```
The comment says 8 * values while this is 12? Seems like a mistype?

sethah replied:
I wondered how confusing this comment might be. Since `values.length == rowIndices.length == numActives`:

`8 * values.length + 4 * rowIndices.length = 8 * numActives + 4 * numActives = 12 * numActives`

The comment is meant to show where each number comes from and the implementation is meant to just be a condensed computation. But please let me know if you think it's too confusing.

MLnick replied:
ah right - no that's fine.

sethah commented:
ping @dbtsai :)

sethah committed:
update since

SparkQA commented:
**[Test build #70135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70135/consoleFull)** for PR 15628 at commit [`10ec41b`](https://github.com/apache/spark/commit/10ec41b4cd07521014ff120cbbf1ecd7545d6d29).

SparkQA commented:
**[Test build #70135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/70135/consoleFull)** for PR 15628 at commit [`10ec41b`](https://github.com/apache/spark/commit/10ec41b4cd07521014ff120cbbf1ecd7545d6d29).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/70135/
Test FAILed.

sethah commented:
pinging other potential reviewers @jkbradley @srowen @MLnick I think this is an important patch for multiclass logistic regression.

sethah commented:
ping @imatiach-msft @dbtsai

sethah commented:
re-ping @dbtsai @MLnick @yanboliang I still think this is an important patch :D

sethah committed:
address review

sethah commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -153,6 +153,86 @@ sealed trait Matrix extends Serializable {
    */
```
I made the change. Not sure if we should do this in a separate PR though.

dbtsai replied:
BTW, it's nice to have return type in public method. Can you add `Unit` as return type?

SparkQA commented:
**[Test build #74280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74280/testReport)** for PR 15628 at commit [`810a077`](https://github.com/apache/spark/commit/810a077b85165d992eb73ae07c1418fbce6724be).

SparkQA commented:
**[Test build #74280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74280/testReport)** for PR 15628 at commit [`810a077`](https://github.com/apache/spark/commit/810a077b85165d992eb73ae07c1418fbce6724be).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74280/
Test FAILed.

sethah commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -153,6 +154,97 @@ sealed trait Matrix extends Serializable {
+   * the current layout order.
+   */
+  @Since("2.2.0")
+  def compressed: Matrix = {
```
It's only a problem if we override/implement it in a subclass. Since it's contained wholly in the trait, it will be fine. I think this is ok to leave, though we could make it final? Also we could make three methods: `compressed`, `compressedCSC`, `compressedCSR`. I think the latter is a good solution, thoughts?

dbtsai replied:
+1 on the later one.

sethah commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -153,6 +154,97 @@ sealed trait Matrix extends Serializable {
+   * @param colMajor Whether the values of the resulting sparse matrix should be in column major
+   *                    or row major order. If `false`, resulting matrix will be row major.
+   */
+  private[ml] def toSparseMatrix(colMajor: Boolean): SparseMatrix
```
You mean `@inline private[ml] def ...` ? Do we expect this to be called often enough for that to make a difference?

dbtsai replied:
Fair enough.

sethah committed:
toDenseRowMajor, toDenseColMajor

SparkQA commented:
**[Test build #74477 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74477/testReport)** for PR 15628 at commit [`cf8945a`](https://github.com/apache/spark/commit/cf8945a83530324701d51fcc212d0becdf4a70c9).

sethah commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -153,6 +154,97 @@ sealed trait Matrix extends Serializable {
+   */
+  @Since("2.2.0")
+  def toDense: DenseMatrix = toDenseMatrix(colMajor = true)
+
```
Added.

sethah commented:
@dbtsai Let me know your thoughts on the comments I left. Thanks for the review!

SparkQA commented:
**[Test build #74477 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74477/testReport)** for PR 15628 at commit [`cf8945a`](https://github.com/apache/spark/commit/cf8945a83530324701d51fcc212d0becdf4a70c9).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74477/
Test FAILed.

sethah committed:
mima

SparkQA commented:
**[Test build #74479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74479/consoleFull)** for PR 15628 at commit [`254b9fb`](https://github.com/apache/spark/commit/254b9fb07a35d6927fefe1a4abe6f8a24ae81d4a).

SparkQA commented:
**[Test build #74486 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74486/testReport)** for PR 15628 at commit [`baa8c9d`](https://github.com/apache/spark/commit/baa8c9daff8e405575c1c733e4001a0c1ccb6796).

SparkQA commented:
**[Test build #74479 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74479/consoleFull)** for PR 15628 at commit [`254b9fb`](https://github.com/apache/spark/commit/254b9fb07a35d6927fefe1a4abe6f8a24ae81d4a).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74479/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74486/
Test FAILed.

sethah commented:
Jenkins test this please.

SparkQA commented:
**[Test build #74552 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74552/testReport)** for PR 15628 at commit [`baa8c9d`](https://github.com/apache/spark/commit/baa8c9daff8e405575c1c733e4001a0c1ccb6796).

SparkQA commented:
**[Test build #74552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74552/testReport)** for PR 15628 at commit [`baa8c9d`](https://github.com/apache/spark/commit/baa8c9daff8e405575c1c733e4001a0c1ccb6796).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74552/
Test PASSed.

sethah commented:
ping @dbtsai! Code freeze is upcoming :)

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,109 @@ sealed trait Matrix extends Serializable {
+   * Converts this matrix to a sparse matrix in column major order.
+   */
+  @Since("2.2.0")
+  def toCSC: SparseMatrix = toSparseMatrix(colMajor = true)
```
I'm not good at naming, but since we use `toDenseRowMajor` for dense vector, should we use `toSparseColumnMajor`? Almost many packages are using `toCSC`, but I think we can make them consistent. Just my 2 cents.

sethah replied:
I'm not sure I have a preference. I don't mind leaving them as CSC and CSR.

dbtsai replied:
After thinking about it again, let's have it as `toSparseColumnMajor` to make the apis consistent with the dense ones if you don't mind?

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,109 @@ sealed trait Matrix extends Serializable {
+   * Converts this matrix to a sparse matrix in row major order.
+   */
+  @Since("2.2.0")
+  def toCSR: SparseMatrix = toSparseMatrix(colMajor = false)
```
Same question.

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,109 @@ sealed trait Matrix extends Serializable {
+   *                    or row major order. If `false`, resulting matrix will be row major.
+   */
+  @Since("2.2.0")
+  def compressed(colMajor: Boolean): Matrix = {
```
Let's make it private, and follow the previous style. Should add `compressedRowMajor`, `compressedColumnMajor` since it can be dense matrix in certain situations.

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -291,31 +395,60 @@ class DenseMatrix @Since("2.0.0") (
+   */
+  private[ml] override def toDenseMatrix(colMajor: Boolean): DenseMatrix = {
+    if (!(isTransposed ^ colMajor)) {
+      val newValues = new Array[Double](numCols * numRows)
```
Simpler to use `foreachActive`? With it, both `toDenseMatrix` can have the same implementation for sparse and dense in trait.

sethah replied:
I don't think we can put this in the trait, since when it is called on dense matrix we would like to return `this` in some cases (when no layout change is needed). But, yes, I think it simpler to use `toArray` which calls `foreachActive`. Thanks!

dbtsai replied:
The following could work, and we only need one implementation in trait. Thanks.

```scala
trait Matrix {

  var isTransposed: Boolean = true
  var numCols: Int = 0
  var numRows: Int = 0

  def foreachActive(f: (Int, Int, Double) => Unit): Unit

  def toDenseMatrix(colMajor: Boolean): Matrix = {
    this match {
      case _: DenseMatrix if this.isTransposed != colMajor =>
        this
      case _: SparseMatrix | _: DenseMatrix if this.isTransposed == colMajor =>
        val newValues = new Array[Double](numCols * numRows)

        this.foreachActive { case (row, col, value) =>
            // filling the newValues
        }
        new DenseMatrix(numRows, numCols, newValues, isTransposed = !colMajor)
      case _ =>
        throw new IllegalArgumentException("")
    }
  }
}

class DenseMatrix extends Matrix {

  def foreachActive(f: (Int, Int, Double) => Unit): Unit = {

  }

}

class SparseMatrix extends Matrix {
  def foreachActive(f: (Int, Int, Double) => Unit): Unit = {

  }
}
```

sethah replied:
Hm, I don't think this solution is better. The entire point of abstract methods is to allow subclasses to implement a method differently. Since we need different implementations depending on the subclass, we should just implement them in the subclasses. We can do this with the following:

````scala
trait Matrix {
  def toDenseMatrix(colMajor: Boolean): Matrix
}
class DenseMatrix extends Matrix {
    private[ml] override def toDenseMatrix(colMajor: Boolean): DenseMatrix = {
    if (isTransposed && colMajor) {
      new DenseMatrix(numRows, numCols, toArray, isTransposed = false)
    } else if (!isTransposed && !colMajor) {
      new DenseMatrix(numRows, numCols, transpose.toArray, isTransposed = true)
    } else {
      this
    }
  }
}
class SparseMatrix extends Matrix {
  private[ml] override def toDenseMatrix(colMajor: Boolean): DenseMatrix = {
    if (colMajor) new DenseMatrix(numRows, numCols, toArray)
    else new DenseMatrix(numRows, numCols, this.transpose.toArray, isTransposed = true)
  }
}
````

Which is less verbose than the previous code. I'm going to put that in the next commit. Let me know what you think.

dbtsai replied:
This looks great to me!

sethah committed:
add compressedColRowMajor

SparkQA commented:
**[Test build #75016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75016/testReport)** for PR 15628 at commit [`17dec63`](https://github.com/apache/spark/commit/17dec637d33c997f724e06ab28487f1e9441394c).

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -587,18 +722,69 @@ class SparseMatrix @Since("2.0.0") (
-    new DenseMatrix(numRows, numCols, toArray)
+  private[ml] override def toSparseMatrix(colMajor: Boolean): SparseMatrix = {
+    if (!(colMajor ^ isTransposed)) {
+      // breeze transpose rearranges values in column major and removes explicit zeros
```
I think it's hacky to use breeze's transpose behavior to remove zeros in sparse matrices. Can we have our own implementation given we're potentially remove breeze?

sethah replied:
I pretty much agree with you, but this is non-trivial code if we want to do it efficiently. Breeze has a pretty well-optimized implementation to do this. I would leave it as a follow up JIRA, or do it when/if we ever remove the Breeze dependency. Or do you think this is a blocker for this PR?

dbtsai replied:
This is not a blocker.

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -587,18 +722,69 @@ class SparseMatrix @Since("2.0.0") (
+        val breezeTransposed = asBreeze.asInstanceOf[BSM[Double]]
+        Matrices.fromBreeze(breezeTransposed).asInstanceOf[SparseMatrix]
+      }
+    } else {
```
Can we document here that it's when the layout of this and colMajor is different? Easier read than `(colMajor ^ isTranspose)` condition here. Even more readable to use pattern matching with exact boolean on both variables.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,385 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(!dm3.isTransposed)
+    assert(dm3.values.equals(dm1.values))
+
+    val dm4 = dm1.toDenseMatrix(false)
```
I would like to make `toDenseMatrix` as private, and we test against `toDenseRowMajor` which is more explicit.

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,110 @@ sealed trait Matrix extends Serializable {
+   * Converts this matrix to a sparse matrix in column major order.
+   */
+  @Since("2.2.0")
+  def toSparse: SparseMatrix = toSparseMatrix(colMajor = true)
```
I'm debating that should we keep the same ordering of layout when we call `toSparse` or `toDense`?

sethah replied:
Well, this would change the behavior of external facing code. Before if I called `toSparse` on a row major matrix, I'd get a column major matrix. If we maintain the layout, then I'd now get something different (column major). Otherwise, I'd agree it is best to maintain the layout.

dbtsai replied:
But I thought this is a new api being added, so we can make it to maintain the same layout.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,385 @@ class MatricesSuite extends SparkMLFunSuite {
+            1.0 -3.0 -8.0
+     */
+    val dm1 = new DenseMatrix(2, 3, Array(4.0, -1.0, 2.0, 7.0, -8.0, 4.0))
+    val dm2 = new DenseMatrix(2, 3, Array(5.0, -9.0, 4.0, 1.0, -3.0, -8.0), isTransposed = true)
```
Why not just make `dm2` dm1.transposed, but explicitly assign the value?  Thus, you don't need to type the value in the array for the comparison.

sethah replied:
I'm not sure I understand your meaning here. These are made to be two entirely different matrices anyway.

SparkQA commented:
**[Test build #75016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75016/testReport)** for PR 15628 at commit [`17dec63`](https://github.com/apache/spark/commit/17dec637d33c997f724e06ab28487f1e9441394c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75016/
Test FAILed.

kiszk commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -291,31 +396,60 @@ class DenseMatrix @Since("2.0.0") (
+  private[ml] override def toDenseMatrix(colMajor: Boolean): DenseMatrix = {
+    if (!(isTransposed ^ colMajor)) {
+      val newValues = new Array[Double](numCols * numRows)
+      var j = 0
```
Can we move `if (isTransposed) {` out of a loop like the following since it is a loop invariant and we want to remove div/mod operations?
```
if (isTransposed) {
  // it is row major and we want column major
  var j = 0
  var col = 0
  while (col < numCols) {
    var row = 0
    while (row < numRows) {
      ...
    }
    col += 1
  }
} else {
  // it is column major and we want row major
  var j = 0
  var row = 0
  while (row < numRows) {
    var col = 0
    while (col < numCols) {
      ...
    }
    rows += 1
  }
}
```

sethah replied:
See above discussion, I am going to change this to use `toArray`.

kiszk replied:
It is fine with me since `isTransposed` is checked beforehand.

kiszk replied:
I have one question regarding performance.
To use [`toArray`](https://github.com/sethah/spark/blob/4746ec0d97c002241be344494a6d2ddee3a7c2d5/mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala#L49-L55) introduce allocation of a temporary array and data copy.
Can we avoid this allocation and copy by passing an original array and access function to `new DenseMatrix`?

sethah replied:
I don't see how there is an unnecessary copy. `toArray` copies the elements of the current matrix to a new Array, then uses that as the backing array of a new `DenseMatrix`. We cannot modify the original matrix values.

kiszk replied:
I see. `DenseMatrix.update()` modifies the contexts of `value`.

sethah committed:
toSparseColMajor

SparkQA commented:
**[Test build #75083 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75083/testReport)** for PR 15628 at commit [`4746ec0`](https://github.com/apache/spark/commit/4746ec0d97c002241be344494a6d2ddee3a7c2d5).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75083/
Test FAILed.

sethah commented:
Jenkins retest this please.

SparkQA commented:
**[Test build #75097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75097/testReport)** for PR 15628 at commit [`4746ec0`](https://github.com/apache/spark/commit/4746ec0d97c002241be344494a6d2ddee3a7c2d5).

SparkQA commented:
**[Test build #75097 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75097/testReport)** for PR 15628 at commit [`4746ec0`](https://github.com/apache/spark/commit/4746ec0d97c002241be344494a6d2ddee3a7c2d5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75097/
Test FAILed.

sethah commented:
Jenkins retest this please.

SparkQA commented:
**[Test build #75116 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75116/testReport)** for PR 15628 at commit [`4746ec0`](https://github.com/apache/spark/commit/4746ec0d97c002241be344494a6d2ddee3a7c2d5).

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,118 @@ sealed trait Matrix extends Serializable {
+      toDenseMatrix(!isTransposed)
+    } else {
+      if (cscSize <= csrSize) {
+        toSparseMatrix(colMajor = true)
```
`toSparseColumnMajor`

sethah replied:
Done

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,118 @@ sealed trait Matrix extends Serializable {
+      if (cscSize <= csrSize) {
+        toSparseMatrix(colMajor = true)
+      } else {
+        toSparseMatrix(colMajor = false)
```
`toSparseRowMajor`

sethah replied:
Done

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -161,6 +162,118 @@ sealed trait Matrix extends Serializable {
+    val csrSize = getSparseSizeInBytes(colMajor = false)
+    if (getDenseSizeInBytes < math.min(cscSize, csrSize)) {
+      // dense matrix size is the same for column major and row major, so maintain current layout
+      toDenseMatrix(!isTransposed)
```
Call `toDense` if we decide to make `toDense` and `toSparse` outputting the same layout.

sethah replied:
Done

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -291,31 +404,49 @@ class DenseMatrix @Since("2.0.0") (
+   * Generate a `DenseMatrix` from this `DenseMatrix`.
+   *
+   * @param colMajor Whether the resulting `DenseMatrix` values will be in column major order.
+   */
```
Minor, can we have 
```scala
protected def isColMajor = !isTransposed

protected def isRowMajor = isTransposed
```

So the code can be understood easier?

sethah replied:
Ok, I added these methods. I updated the test suites to use them instead of `isTransposed`.

sethah committed:
toSparse, toDense maintain current layout

SparkQA commented:
**[Test build #75122 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75122/testReport)** for PR 15628 at commit [`354935b`](https://github.com/apache/spark/commit/354935b2b89c3d614b905ed140727c4786c40978).

SparkQA commented:
**[Test build #75123 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75123/testReport)** for PR 15628 at commit [`4026e89`](https://github.com/apache/spark/commit/4026e892c60a0c1866fd61fc5137c4b5d1bcb13b).

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -291,31 +404,49 @@ class DenseMatrix @Since("2.0.0") (
     }
-    new SparseMatrix(numRows, numCols, colPtrs, rowIndices.result(), spVals.result())
   }
 
```
Could we override the `toArray` in DenseMatrix so when `this` is column major, we just return `this.values`? Otherwise, it's very expensive to create a new array.

sethah replied:
I have to say I'm a bit surprised - I thought that's what `toArray` already did! Yeah, that seems like a good change, but I'd prefer to do it in another pr because we need to make sure that this doesn't adversely affect other places that use `toArray` as well as adding unit tests. If that sounds ok, I'll make a JIRA for it.

SparkQA commented:
**[Test build #75116 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75116/testReport)** for PR 15628 at commit [`4746ec0`](https://github.com/apache/spark/commit/4746ec0d97c002241be344494a6d2ddee3a7c2d5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75116/
Test PASSed.

SparkQA commented:
**[Test build #75123 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75123/testReport)** for PR 15628 at commit [`4026e89`](https://github.com/apache/spark/commit/4026e892c60a0c1866fd61fc5137c4b5d1bcb13b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75123/
Test PASSed.

SparkQA commented:
**[Test build #75122 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75122/testReport)** for PR 15628 at commit [`354935b`](https://github.com/apache/spark/commit/354935b2b89c3d614b905ed140727c4786c40978).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75122/
Test PASSed.

sethah committed:
add isRowMajor, isColMajor

SparkQA commented:
**[Test build #75144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75144/testReport)** for PR 15628 at commit [`c880a45`](https://github.com/apache/spark/commit/c880a45f2ca877d8f228ca3147049b5b5ebbd43d).

SparkQA commented:
**[Test build #75144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75144/testReport)** for PR 15628 at commit [`c880a45`](https://github.com/apache/spark/commit/c880a45f2ca877d8f228ca3147049b5b5ebbd43d).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75144/
Test FAILed.

SparkQA commented:
**[Test build #75147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75147/testReport)** for PR 15628 at commit [`93ec250`](https://github.com/apache/spark/commit/93ec250b2b0268f57e2c4d191f27da22a74d3f0c).

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val dm3 = dm1.toDense
+    assert(dm3 === dm1)
+    assert(!dm3.isTransposed)
+    assert(dm3.values.equals(dm1.values))
```
`val dm4 = dm1.toDenseRowMajor` and `val dm7 = dm1.toDenseRowMajor` are the same.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+      Array(4.0, 5.0, 2.0, 0.0), isTransposed = true)
+    val smZeros = new SparseMatrix(2, 3, Array(0, 2, 4, 6), Array(0, 1, 0, 1, 0, 1),
+      Array(0.0, 0.0, 0.0, 0.0, 0.0, 0.0))
+
```
Group the tests by methods

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(!dm6.isTransposed)
+    assert(dm6.values === Array(0.0, 0.0, 4.0, 2.0, 5.0, 0.0))
+
+    val dm7 = sm2.toDenseRowMajor
```
This is duplicated to `val dm4 = sm2.toDenseRowMajor`. And some of them don't have test of `toDenseColMajor`

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val cm1 = dm1.compressed.asInstanceOf[SparseMatrix]
+    assert(cm1 === dm1)
+    assert(cm1.isTransposed)
+    assert(cm1.getSizeInBytes <= dm1.getSizeInBytes)
```
Why not `<` instead of `<=`?

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val cm2 = dm1.compressedColMajor.asInstanceOf[SparseMatrix]
+    assert(cm2 === dm1)
+    assert(!cm2.isTransposed)
+    assert(cm2.getSizeInBytes <= dm1.getSizeInBytes)
```
ditto

dbtsai replied:
Also check compressedRowMajor.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val cm3 = dm2.compressed.asInstanceOf[SparseMatrix]
+    assert(cm3 === dm2)
+    assert(!cm3.isTransposed)
+    assert(cm3.getSizeInBytes <= dm2.getSizeInBytes)
```
ditto.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(cm4 === dm3)
+    assert(!cm4.isTransposed)
+    assert(cm4.values.equals(dm3.values))
+    assert(cm4.getSizeInBytes <= dm3.getSizeInBytes)
```
Should be `=` here.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val cm5 = dm3.compressedRowMajor.asInstanceOf[DenseMatrix]
+    assert(cm5 === dm3)
+    assert(cm5.isTransposed)
+    assert(cm5.getSizeInBytes <= dm3.getSizeInBytes)
```
Should be `=` here.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(cm6 === dm4)
+    assert(cm6.isTransposed)
+    assert(cm6.values.equals(dm4.values))
+    assert(cm6.getSizeInBytes <= dm4.getSizeInBytes)
```
ditto.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(cm1 === sm1)
+    assert(!cm1.isTransposed)
+    assert(cm1.values.equals(sm1.values))
+    assert(cm1.getSizeInBytes <= sm1.getSizeInBytes)
```
ditto. remove `=`

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(cm2.isTransposed)
+    // forced to be row major, so we have increased the size
+    assert(cm2.getSizeInBytes > sm1.getSizeInBytes)
+    assert(cm2.getSizeInBytes <= sm1.toDense.getSizeInBytes)
```
`=` is needed?

sethah replied:
I think it's ok to leave it here since they could potentially be equal.

dbtsai replied:
If =, dense should be preferred since it will be faster. As a result, we should only check <.

sethah replied:
Ok, before we broke ties with sparse, so I changed it to choose dense, and added unit tests. Also removed the = here.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    assert(cm8 === sm2)
+    assert(!cm8.isTransposed)
+    assert(cm8.getSizeInBytes > sm2.getSizeInBytes)
+    assert(cm8.getSizeInBytes <= sm2.toDense.getSizeInBytes)
```
For completion, could you test `sm1. compressedColumnMajor` and `sm2. compressedRowMajor`? Thanks.

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val cm4 = sm3.compressed.asInstanceOf[DenseMatrix]
+    assert(cm4 === sm3)
+    assert(!cm4.isTransposed)
+    assert(cm4.getSizeInBytes <= sm3.getSizeInBytes)
```
remove `=`?

dbtsai commented on mllib-local/src/test/scala/org/apache/spark/ml/linalg/MatricesSuite.scala:
```diff
@@ -160,22 +160,395 @@ class MatricesSuite extends SparkMLFunSuite {
+    val cm5 = sm3.compressedRowMajor.asInstanceOf[DenseMatrix]
+    assert(cm5 === sm3)
+    assert(cm5.isTransposed)
+    assert(cm5.getSizeInBytes <= sm3.getSizeInBytes)
```
remove `=`? and `sm3.compressedColumnMajor`?

sethah committed:
organize test suites

SparkQA commented:
**[Test build #75155 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75155/testReport)** for PR 15628 at commit [`5411d46`](https://github.com/apache/spark/commit/5411d4629da1dfa7cc4db83a2741c4f80cbd7bf3).

sethah commented:
@dbtsai Thanks for the good suggestions. I rearranged the test suites, removed redundancies, and filled in some gaps. Things got a bit jumbled when changing some of the methods around. I also added the `this` prefix to some of the method calls as you suggest.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75155/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75147/
Test FAILed.

dbtsai reviewed (commented):
Minor

sethah committed:
break ties with dense

SparkQA commented:
**[Test build #75167 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75167/testReport)** for PR 15628 at commit [`95ac0e0`](https://github.com/apache/spark/commit/95ac0e0e6ad43646477ef8a359b6dd977764c455).

dbtsai commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -1079,4 +1267,15 @@ object Matrices {
+
+  private[ml] def getSparseSize(numActives: Long, numPtrs: Long): Long = {
+    // 8 * values.length + 4 * rowIndices.length + 4 * colPtrs.length + 12 + 12 + 12 + 1
+    12L * numActives + 4L * numPtrs + 37L
```
`(java.lang.Double.BYTES + java.lang.Integer.BYTES) * numActives + java.lang.Integer.BYTES * numPtrs + 37L`

dbtsai replied:
Nice that we can get `37L` using java apis to ensure the portability.

dbtsai reviewed (approved):
LGTM. Couple small comments.

sethah committed:
clarify get size functions

SparkQA commented:
**[Test build #75171 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75171/testReport)** for PR 15628 at commit [`87dfaa0`](https://github.com/apache/spark/commit/87dfaa06b2861ccc4bd5ca62915dea867985b391).

SparkQA commented:
**[Test build #75167 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75167/testReport)** for PR 15628 at commit [`95ac0e0`](https://github.com/apache/spark/commit/95ac0e0e6ad43646477ef8a359b6dd977764c455).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75167/
Test PASSed.

sethah commented:
Thanks for finally cooperating, Jenkins!

dbtsai commented:
Thanks @sethah and Jenkins! Merged into master.

sethah commented:
Thanks for all the time reviewing!

SparkQA commented:
**[Test build #75171 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/75171/testReport)** for PR 15628 at commit [`87dfaa0`](https://github.com/apache/spark/commit/87dfaa06b2861ccc4bd5ca62915dea867985b391).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/75171/
Test PASSed.

jkbradley commented on mllib-local/src/main/scala/org/apache/spark/ml/linalg/Matrices.scala:
```diff
@@ -148,7 +154,8 @@ sealed trait Matrix extends Serializable {
    */
-  private[spark] def foreachActive(f: (Int, Int, Double) => Unit)
+  @Since("2.2.0")
+  def foreachActive(f: (Int, Int, Double) => Unit): Unit
```
@sethah @dbtsai Hi all, just saw this during QA.  This method is not very Java-friendly.  I'm OK with adding it as long as we document the fact that it's not Java-friendly.  We could also consider adding a Java-friendly version, perhaps using https://spark.apache.org/docs/latest/api/java/org/apache/spark/api/java/function/Function2.html

dbtsai replied:
I think we should do the same for `foreachActive` in Vector which was already a public api long time ago.

jkbradley replied:
That's a good point.  I guess we can leave it until someone complains & add a Java-friendly one as needed.
